### PR TITLE
Support for SCRAM-SHA-256 SASL authentication

### PR DESCRIPTION
**Pull request by gwynne**
- PostgreSQL supports `SCRAM-SHA-256` authentication since version 11. This is preliminary support for that authentication type.

### CODE REVIEW

**Comment by gwynne** [comment]
> One question though, should there be tests for this?

@calebkleveter The Postgres-specific side of it is implicitly tested whenever the test suite is run against a database configured for `scram-sha-256` auth, a job for a Docker Compose manifest or similar. But yes, there certainly should be unit tests for the other bits. Some categories that'd make good test suites that come to mind:
- Tests for the generic SASL authentication manager (such as by implementing a dummy mechanism to put the system through its paces)
- Tests for the `SCRAM_SHA256` mechanism in terms of it as a `SASLMechanism`.
- Tests for the `SCRAMMessageParser`, which is currently a bit of a hacked-up mess with several fail-closed (auth fails when it shouldn't rather than vice versa) corner cases.
- Possibly, separate tests for the `GS2Header` parsing code, depending how later refinements to the overall implementation go.
- Tests for the `Hi()` function, which is effectively most of a PBKDF2 implementation and should at least be verified against that algorithm's test vectors. 
- Last but not least, tests to verify the behavior of the additional as-yet-unimplemented `SCRAM_SHA256_PLUS` mechanism, particularly to verify it performs channel bonding  correctly, since that's the only reason it exists.

File: Sources/PostgresNIO/Connection/PostgresConnection+Authenticate.swift [hunk 1]
```diff
@@ -60,12 +63,64 @@ private final class PostgresAuthenticationRequest: PostgresRequest {
...
             default: throw PostgresError.protocol("Unexpected response to start message: \(message)")
             }
+        case .saslInitialSent(let manager),
+             .saslChallengeResponse(let manager):
+            switch message.identifier {
+            case .authentication:
+                let auth = try PostgresMessage.Authentication(message: message)
+                switch auth {
+                case .saslContinue(let data), .saslFinal(let data):
+                    print(auth)
```
Comment (tanner0101) [inline Sources/PostgresNIO/Connection/PostgresConnection+Authenticate.swift:98]: Rm `print`.

File: Sources/PostgresNIO/Utilities/SASLAuthentication+SCRAM-SHA256.swift [hunk 2]
```diff
@@ -0,0 +1,658 @@
...
+}
+
+
+/// Implementation of `SCRAM-SHA-256` as a `SASLAuthenticationMechanism`
+///
+/// Implements SCRAM-SHA-256 as described by:
+/// - [RFC 7677 (SCRAM-SHA-256 and SCRAM-SHA-256-PLUS SASL Mechanisms)](https://tools.ietf.org/html/rfc7677)
+/// - [RFC 5802 (SCRAM SASL and GSS-API Mechanisms)](https://tools.ietf.org/html/rfc5802)
+/// - [RFC 4422 (Simple Authentication and Security Layer)](https://tools.ietf.org/html/rfc4422)
+public struct SASLMechanism_SCRAM_SHA256: SASLAuthenticationMechanism {
```
Comment (tanner0101) [inline Sources/PostgresNIO/Utilities/SASLAuthentication+SCRAM-SHA256.swift:285]: Maybe `SASLMechanismSCRAMSHA256` or `SASLMechanism.SCRAM.SHA256` ?

File: Sources/PostgresNIO/Connection/PostgresConnection+Authenticate.swift [hunk 3]
```diff
@@ -60,12 +63,64 @@ private final class PostgresAuthenticationRequest: PostgresRequest {
                     return try [PostgresMessage.Password(string: hash).message()]
                 case .plaintext:
                     return try [PostgresMessage.Password(string: self.password ?? "").message()]
+                case .saslMechanisms(let saslMechanisms):
+                    if saslMechanisms.contains("SCRAM-SHA-256") && self.password != nil {
+                        let saslManager = SASLAuthenticationManager(asClientSpeaking:
+                            SASLMechanism_SCRAM_SHA256(username: self.username, password: { self.password! }))
```
Comment (tanner0101) [inline Sources/PostgresNIO/Connection/PostgresConnection+Authenticate.swift:69]: If this is all handled implicitly, does `SASLMechanism_SCRAM_SHA256` and friends need to be public?

Comment (gwynne) [reply Sources/PostgresNIO/Connection/PostgresConnection+Authenticate.swift:69]: It was originally intended for these to be more globally useful, not limited to Postgres. I suppose they should be kept internal until such time as they actually _are_ ready for such use though.

**Comment by gwynne** [comment]
@tanner0101 Test workflow has been updated to test all three auth methods (trust, MD5, SCRAM-SHA-256) for all supported PostgreSQL versions (and lots of Swift versions...), `docker-compose.yml` now permits specifying an auth method via env var. I'd like to get this in ASAP; more Postgres setups are starting to default to SCRAM-SHA-256, and Postgres itself has made it the default for the "next version after 13" release (see https://www.postgresql.org/docs/devel/runtime-config-connection.html#RUNTIME-CONFIG-CONNECTION-AUTHENTICATION, at the time of this writing).

**Comment by bridger** [comment]
Thank you for implementing this! I was getting my dev environment set up on a new machine and sure enough, postgresql has started defaulting to scram-sha-256. That means it was rejecting any connections from postgres-nio. I'm happy to see the fix is ready to go! 🌟

**Review by Joannis — APPROVED** [review]
A lot of commented code could be removed, but it's non-critical for sure. LGTM

**Comment by tanner0101** [comment]
These changes are now available in [1.4.2](https://github.com/vapor/postgres-nio/releases/tag/1.4.2)
